feat(metro-config): Implement standalone live-binding import-export plugin

kitten opened the pull request:
Resolves #39201
Resolves #39277

## Why

We've run into some limitations with the current implementation of the `importExportPlugin`. Currently, this plugin is closely related and derived from Metro's implementation and dynamically switches to a live-binding output. This approach won't stop us from implementing a fully-compliant version of our ESM transpilation, however, it currently serves two different outputs.

@krystofwoldrich and I previously talked about some possible changes that would be possible, if we started without the constraints of the original `import-export-plugin`, and some tweaks to the output we could make. This PR is an implementation of (most) of these ideas.

Notable changes in requirements are:
- We'd like to consistently drop imports when they're unused (excluding side-effect imports)
- We can "invert" the import specifier access, just like Rollup does in all cases (even defaults) to simplify the logic and reduce special cases
- We can remove our usage of `_$$_IMPORT_DEFAULT` and `_$$_IMPORT_ALL`, insetad deduplicating require-calls aggressively to a single one per source

A few bugs that are fixed that we wanted to / can address in the other implementation too:
- Declarations hoisted out of `export default`s shouldn't conflict with other names in scope
- Re-export-all declarations shouldn't hoisted into exports, but should instead be pushed down to the imports section
- Live bindings should be applied to variable declarations that are exported
- Exports must be created for nested variable declaration patterns
- Namespace names (or other exports) shouldn't conflict with local names (when using `as` aliasing)

## How

The static import-export plugin has been separated from the full `import-export-plugin`. The prior implementation for live bindings is being replaced with a new implementation in this PR.

Roughly, this implementation works with the following phases:
1. Scan imports and register imports, keeping track of requires, default imports, side-effect imports, and namespace-imports. This sets up usage tracking (while default/namespace imports depend on a require being added) and it sets up renames ("inline body refs")
2. Export-all declarations are tracked separately and added to a list of export-alls, while registering an import and marking it as used
3. Export-default declarations are hoisted out of the export and added to a list of "deferred exports" to be processed later
4. Named exports without a source (local exports) are added to a list of "deferred exports" to be processed later
5. Named exports with a source are registered, registering requires/default-imports/namespace-imports for them, and marking them as used. An export is added to a list of exports for them
6. We process all "deferred exports" adding them to the list of exports while renaming their local variables (with "inline body refs") as needed
7. We traverse all remaining local variables that are tracked in the list of "inline body refs" and apply renames, marking imports as used as we find usage sites
8. We make sure all imports that are used in exports are marked as used
9. We add exports to the top of the program (export for export-all declarations)
10. We add imports after to the top of the program (adding export-all declarations after their corresponding import)
11. We add the `isESModule` marker, if needed

# Test Plan

This PR also includes a new test suite (one of two new ones we have so far) that evaluates test cases (from Babel) and checks them against expected outputs. This imitates CommonJS and checks for both export values and import order.

Old tests have been updated, output has been validated manually.

A separate PR will supplement snapshot tests based on Babel's test cases.

We're also planning to:
- [x] Test against apps (and previously failing reproductions, or older reproductions) manually
- [x] Try out common large libraries that have complex ESM structures

# Checklist

<!--
Please check the appropriate items below if they apply to your diff.
-->

- [x] I added a `changelog.md` entry and rebuilt the package sources according to [this short guide](https://github.com/expo/expo/blob/main/CONTRIBUTING.md#-before-submitting)
- [ ] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)

kitten committed:
Add standalone live bindings import-export implementation based on Rollup's output

kitten committed:
Move import-export to importExportStatic and remove live binding section

kitten committed:
Switch between plugins based on liveBindings option

kitten committed:
Number all snapshot test cases

kitten committed:
Improve reusing other unused local IDs

When we're creating a default or namespace import, we can reuse
the name safely, if it's not been assigned a string-name

kitten committed:
Fix isESModule setting: We must set this even if we only have imports

kitten committed:
Skip unreferenced default/namespace imports

kitten committed:
Add source locations via withLocation to importExportLiveBindings

kitten committed:
Update tests for new structure

kitten committed:
Update importExport-snapshots

kitten committed:
Restore side-effect import not having variable declaration

kitten committed:
Stop hoisting export-all declarations which breaks import-order

While it seems intuitively correct to make exports hoist corresponding
to exports coming first, they're part of the import statement block.

kitten committed:
Add snapshot test for nested destructure

kitten committed:
Fix import dedupe

kitten committed:
Add evaluated test suite

kitten committed:
Remove redundant ImportRef.loc and side-effect import on empty export-from

kitten committed:
Update remaining tests

kitten committed:
Fix isESModule override in importExportStatic

kitten committed:
Rebuild

kitten committed:
Add CHANGELOG entry

krystofwoldrich commented on packages/@expo/metro-config/src/transform-plugins/helpers.ts:
```diff
@@ -0,0 +1,167 @@
+    t.variableDeclarator(t.identifier(id), t.callExpression(t.identifier('require'), [source])),
+  ]);
+
+/** `var %id% = require("%source%");` */
```
````suggestion
/** `require("%source%");` */
````

kitten committed:
Update packages/@expo/metro-config/src/transform-plugins/helpers.ts

Co-authored-by: Krystof Woldrich <31292499+krystofwoldrich@users.noreply.github.com>

kitten committed:
Clarify side-effects / explicitly dedup

krystofwoldrich committed:
[metro-config] Merge renameTopLevelModuleVariables into importExportLiveBindings plugin (#39321)

krystofwoldrich committed:
[metro-config] Move unused elimination from Import Export Plugin into Tree Shaking Plugin (#39351)

kitten committed:
Add special handling for `__proto__` edge case

kitten committed:
test(metro-config): Add tests for past import-export-plugin changes (#39305)

Merging into #39282 

Adds initial tests for prior failing behaviour based on current/past
issues and changes/PRs.

krystofwoldrich committed:
Add Babel's `transform-modules-commonjs` test suit for `import-export-plugin` (#39318)

krystofwoldrich committed:
Return export-expr-with-same-name babel snapshot test
